source/base/hdf5.cc

```diff
@@ -0,0 +1,1752 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_HDF5
+
+#  include <deal.II/base/hdf5.h>
+
+#  include <deal.II/lac/full_matrix.h>
+#  include <deal.II/lac/vector.h>
+
+#  include <hdf5.h>
+
+#  include <memory>
+#  include <numeric>
+#  include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace HDF5
+{
+  namespace internal
+  {
+    // This function gives the HDF5 datatype corresponding to the C++ type. In
+    // the case of std::complex types the HDF5 handlers are automatically freed
+    // using the destructor of std::shared_ptr.
+    // std::shared_ptr is used instead of std::unique_ptr because the destructor
+    // of std::shared_ptr doesn't have to be defined in the template argument.
+    // In the other hand, the destructor of std::unique has to be defined in the
+    // template argument. Native types such as H5T_NATIVE_DOUBLE does not
+    // require a destructor, but compound types such as std::complex<double>
+    // require a destructor to free the HDF5 resources.
+    template <typename number>
+    std::shared_ptr<hid_t>
+    get_hdf5_datatype()
+    {
+      std::shared_ptr<hid_t> t_type;
+      if (std::is_same<number, float>::value)
+        {
+          return std::make_shared<hid_t>(H5T_NATIVE_FLOAT);
+        }
+      else if (std::is_same<number, double>::value)
+        {
+          return std::make_shared<hid_t>(H5T_NATIVE_DOUBLE);
+        }
+      else if (std::is_same<number, int>::value)
+        {
+          return std::make_shared<hid_t>(H5T_NATIVE_INT);
+        }
+      else if (std::is_same<number, unsigned int>::value)
+        {
+          return std::make_shared<hid_t>(H5T_NATIVE_UINT);
+        }
+      else if (std::is_same<number, std::complex<float>>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
```

Comment:
I think this is a C++14 feature: `[](auto pointer)`. We can't use that yet. Can you specify the argument type explicitly?

Reply:
Yes, indeed: https://en.wikipedia.org/wiki/C%2B%2B14#Generic_lambdas